src/charm.py

```diff
@@ -642,580 +107,251 @@ def k8s_auth(self) -> None:
-            InvalidHostnameError: if there is an invalid hostname.
-            InvalidBackendProtocolError: if an invalid backend protocol is used.
-        """
-        msg = ""
-        # We only want to do anything here if we're the leader to avoid
-        # collision if we've scaled out this application.
-        svc_names = [conf_or_rel._service_name for conf_or_rel in self._all_config_or_relations]
-        if self.unit.is_leader() and any(svc_names):
-            try:
-                self._define_services()
-                self._define_ingresses()
-                msgs = []
-                self.unit.status = WaitingStatus("Waiting for ingress IP availability")
-                ingress_ips = self._report_ingress_ips()
-                if ingress_ips:
-                    msgs.append(f"Ingress IP(s): {', '.join(ingress_ips)}")
-                msgs.append(f"Service IP(s): {', '.join(self._report_service_ips())}")
-                msg = ", ".join(msgs)
-                self._delete_unused_resources()
-            except kubernetes.client.exceptions.ApiException as exception:
-                if exception.status == 403:
-                    LOGGER.error(
-                        "Insufficient permissions to create the k8s service, "
-                        "will request `juju trust` to be run"
-                    )
-                    juju_trust_cmd = f"juju trust {self.app.name} --scope=cluster"
-                    self.unit.status = BlockedStatus(
-                        f"Insufficient permissions, try: `{juju_trust_cmd}`"
-                    )
-                    return
-                raise
-            except ConflictingAnnotationsError:
-                self.unit.status = BlockedStatus(
-                    "Conflicting annotations from relations. Run juju debug-log for details. "
-                    "Set manually via juju config."
-                )
+            ApiException: if kubernetes API error happens.
+        """
+        self.unit.set_workload_version(kubernetes.__version__)
+        try:
+            self._check_precondition()
+            relation = self._get_relation()
+            if relation is None:
+                self._cleanup()
+                self.unit.status = WaitingStatus("waiting for relation")
                 return
-            except ConflictingRoutesError:
+            definition = self._get_definition_from_relation(relation)
+            ingress = self._reconcile(definition)
+            self.unit.status = WaitingStatus("Waiting for ingress IP availability")
+            ingress_ips = self._report_ingress_ips(ingress)
+            message = f"Ingress IP(s): {', '.join(ingress_ips)}" if ingress_ips else ""
+            self.unit.status = ActiveStatus(message)
+        except InvalidIngressError as exc:
+            self.unit.status = BlockedStatus(exc.msg)
+        except kubernetes.client.exceptions.ApiException as exception:
```

Comment:
It would be better to handle a more specific exception here, the k8s client is not used in this function and it raises a k8s exception which is confusing

Reply:
Okay, I added a decorator in `controller.py` to remap exceptions.